File: backend/search-scrapper/amazon-search.js
const puppeteer = require('puppeteer-extra');
const { findMatch } = require("../utils/search-match.js");
const StealthPlugin = require('puppeteer-extra-plugin-stealth');

puppeteer.use(StealthPlugin());

const AmazonSearchProduct = async (req, res) => {
    try {
        const search_query = req.query.q;
        if (!search_query) {
            return res.status(400).json({ error: 'No search query provided' });
        }

        const browser = await puppeteer.launch({
            headless: true,
            args: ['--no-sandbox', '--disable-setuid-sandbox'],
        });
        const page = await browser.newPage();

        await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.71 Safari/537.36');

        await page.setViewport({ width: 1366, height: 768 });

        await page.goto(`https://www.amazon.in/s?k=${search_query}`, { waitUntil: 'domcontentloaded' });

        // await autoScroll(page);

        const products = await page.evaluate(() => {
            const data = [];
            const titles = [];

            document.querySelectorAll('.puis-card-container').forEach(product => {
                if (product) {
                    const title = product.querySelector('.a-size-medium')?.innerText.trim();
                    const discount_price = parseInt(product.querySelector('.a-price-whole')?.innerText.trim().replace(/[^\d.]/g, ''));
                    const original_price = parseInt(product.querySelector('.a-price.a-text-price')?.firstChild?.innerText.trim().replace(/[^\d.]/g, ''));
                    const link = 'https://www.amazon.in' + product.querySelector('.a-link-normal')?.getAttribute('href');
                    const image = product.querySelector('.s-image')?.getAttribute('src');
                    const discount = Math.round(((original_price - discount_price) / original_price) * 100);
                    if (title && discount_price && original_price && discount && link && image) {
                        data.push({ title, discount_price: parseInt(discount_price), original_price: parseInt(original_price), discount: parseInt(discount), link, image });
                        titles.push(title);
                    }
                }
            });

            return { data, titles };
        });

        await browser.close();

        if (products.data.length > 0) {
            // const product_index = findMatch(search_query, products.titles);
            // if (product_index !== null) {
            //     const bestMatch = products.data[product_index];
            //     const otherProducts = products.data.filter((_, index) => index !== product_index);
            //     return res.json({
            //         best: bestMatch,
            //         data: otherProducts
            //     });
            // }
            const otherProducts = products.data.splice(1)
            return res.json({
                best: products.data[0],
                data: otherProducts
            });
        }
        return res.status(404).json({ error: 'No matching product found' });
    } catch (error) {
        console.log("Get Amazon Search Product Error: ", error);
        return res.status(500).json({ error: 'Internal server error' });
    }
};

module.exports = AmazonSearchProduct;

async function autoScroll(page) {
    await page.evaluate(async () => {
        await new Promise((resolve, reject) => {
            var totalHeight = 0;
            var distance = 100;
            var timer = setInterval(() => {
                var scrollHeight = document.body.scrollHeight;
                window.scrollBy(0, distance);
                totalHeight += distance;
                if (totalHeight >= scrollHeight) {
                    clearInterval(timer);
                    resolve();
                }
            }, 100);
        });
    });
}

File: search-scrapper-python/amazon_search.py
from flask import Flask, request, jsonify
import requests
from bs4 import BeautifulSoup
from utils.search_match import find_match

app = Flask(__name__)


def amazon_search_product(search_query):
    try:
        if not search_query:
            return {"error": "No search query provided"}, 400

        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3'
        }

        url = f"https://www.amazon.in/s?k={search_query}"
        response = requests.get(url, headers=headers)
        if response.status_code == 200:
            soup = BeautifulSoup(response.content, 'html.parser')

            products = []
            titles = []

            product_containers = soup.find_all('div', class_='s-result-item')
            for product in product_containers:
                title = product.find('span', class_='a-text-normal')
                discount_price = product.find('span', class_='a-offscreen')
                original_price = product.find('span', class_='a-price')
                link_tag = product.find('a', class_='a-link-normal', href=True)
                image = product.find('img', class_='s-image')

                if title and discount_price and original_price and link_tag and image:
                    title_text = title.get_text(strip=True)
                    discount_price_text = int(discount_price.get_text(
                        strip=True).replace(',', '').replace('₹', ''))
                    original_price_text = int(original_price.find(
                        'span', class_='a-offscreen').get_text(strip=True).replace(',', '').replace('₹', ''))
                    discount = int(
                        round(((original_price_text - discount_price_text) / original_price_text) * 100))
                    link = 'https://www.amazon.in' + link_tag['href']
                    image_src = image['src']

                    products.append({
                        'title': title_text,
                        'discount_price': discount_price_text,
                        'original_price': original_price_text,
                        'discount': discount,
                        'link': link,
                        'image': image_src
                    })
                    titles.append(title_text)

            if products:
                # Match the best product
                product_index = find_match(search_query, titles)
                if product_index is not None:
                    best_match = products[product_index]
                    other_products = [product for i, product in enumerate(
                        products) if i != product_index]
                    return {"best": best_match, "data": other_products}

        return {"error": "No matching product found"}, 404

    except Exception as e:
        print("Get Amazon Search Product Error: ", e)
        return {"error": "An error occurred while processing the request"}, 500


@app.route('/amazon/search')
def search_amazon_product():
    search_query = request.args.get('q')
    result = amazon_search_product(search_query)
    return jsonify(result)


if __name__ == '__main__':
    app.run(debug=True)
